---
output: 
  stevetemplates::article:
    fig_caption: true
#bibliography: master.bib
biblio-style: apsr
title: "phruta: scrapping genbank and assembling phylogenetic trees"
thanks: "Replication files are available on the author's Github account (http://github.com/cromanpa). **Current version**: `r format(Sys.time(), '%B %d, %Y')`; **Corresponding author**: cromanpa94@arizona.edu."
author:
- name: Cristian Rom√°n-Palacios
  affiliation: School of Information, University of Arizona, Tucson, Arizona 85721, USA. ORCiD`:` 0000-0003-1696-4886
abstract: "Current methodological practices for assembling phylogenetic trees often recur to sequence data stored in GenBank. However, understanding molecular and taxonomic availability in GenBank is generally not very straightforward. For instance, the genetic makeup of datasets available in GenBank can strongly differ between genera even within the same family. Similarly, the taxonomic information associated with sequence data in GenBank can be outdated, relative to other databases that mainly focus on the taxonomic side. phruta, a newly developed R package, is designed to improve the user experience and access to information to genetic data stored in GenBank. By using phruta, users are able to (1) quantitatively explore the molecular makeup of particular clades with information in GenBank, (2) assemble curated multi-gene molecular datasets with retrieved and local sequences, and (3) run basic phylogenetic talks, all within R. The structure of the functions implemented in phruta, designed as a workflow, aim to allow users to assemble simple workflows for particular talks, which are in turn expected to increase reproducibility when assembling phylogenies. This paper provides a brief overview on the performance and workflow associated with phruta. "
keywords: "R package, Phylogenetics, Reproducibility, Workflow"
date: "`r format(Sys.time(), '%B %d, %Y')`"
geometry: margin=1in
fontfamily: mathpazo
fontsize: 12pt
spacing: single
endnote: no
---

```{r setup, include=FALSE}
library(knitr)
library(kableExtra)
library(benchmarkme)
library(ggtree)
knitr::opts_chunk$set(cache = TRUE,
                      message = FALSE, 
                      warning = FALSE,
                      fig.path = 'figs/',
                      cache.path = '_cache/',
                      fig.process = function(x) {
                      x2 = sub('-\\d+([.][a-z]+)$', '\\1', x)
                      if (file.rename(x, x2)) x2 else x
                      },
                      tidy.opts = list(width.cutoff = 60), tidy = TRUE)
```


## Background

## The `phruta` `R` package

The `phruta` package is designed to simplify the basic phylogenetic pipeline in `R`. `phruta` is expected to allow scientists from different backgrounds to assemble molecular databases or phylogenies for particular taxa with as minimal complexity and maximal reproducibility as possible. All the code in `phruta` is run within the same software (`R`) and data from intermediate steps are either stored to the environment or can be exported locally to different folders. In general, `phruta` is able to (1) find potentially (phylogenetically) relevant gene regions for a given set of taxa based on GenBank, (2) retrieve gene sequences and curate taxonomic information from the same database, (3) combine downloaded and local gene sequences, and (4) perform sequence alignment, phylogenetic inference, and basic tree dating tasks.

## Alternatives to phruta

Similar functionalities for assembling curated molecular datasets for phylogenetic analyses can be found in [`phylotaR`](https://github.com/ropensci/phylotaR) and [SuperCRUNCH](https://github.com/dportik/SuperCRUNCH). However, `phylotaR` is limited to downloading and curating sequences (e.g. it does not align sequences). Similarly, `SuperCRUNCH` only curates sequences that are already stored locally. In fact, `phruta` is closer to [`SUPERSMART`](https://academic.oup.com/sysbio/article/66/2/152/2418028) and the associated `R` workflow [`SUPERSMARTR`](https://github.com/AntonelliLab/supersmartR). However, most of the applications in the different packages that are part of `SUPERSMARTR` are simplified in `phruta`. Standalone applications that might resemble `phruta` could include `MEGA` and geneious. However, analyses in these two alternatives are either poorly reproducible (e.g. `MEGA`) or not all the functions are freely available  to everyone (e.g. geneious has a paid version). 

## `phruta` in a nutshell

The current release of `phruta` includes a set of eight major functions. Running all eight major functions in `phruta` results in a time-calibrated phylogeny. However, users interested in using their own files at any stage can run each function independently. Note that all the functions for which their primary output are sequences (aligned or unaligned) are listed under `sq.*`. All the files that output phylogenies (time-calibrated or not) are listed under `tree.*`.

-   First, the distribution of gene sampled for a given organism or set of taxa can be explored using the `acc.gene.sampling` function. This function will return a table that summarizes either the distribution of genes sampled for the search term in general or specifically across species.

-   Second, given a list of target organisms, users can retrieve a list of accession numbers that are relevant to their search using `acc.table.retrieve()`. Instead of directly downloading sequences from genbank (see `sq.retrieve.direct()` below), retrieving accession numbers allow users to have more control over the sequences that are being used in the analyses. Note that users can also curate the content of the dataset obtained using `sq.retrieve.direct()`.

-   Third, users should download gene sequences. Sequences can be download using the `sq.retrieve.indirect()` from the accession numbers retrieved before using the `acc.table.retrieve()` function. This is the preferred option within `phruta`. Additionally, users can directly download gene sequences using the `sq.retrieve.direct()` function. Both `sq.retrieve.indirect()` and `sq.retrieve.direct()` functions save gene sequences in `fasta` files that will be located in a new directory named `0.Sequences`.

-   Fourth, `sq.add()` allows users to include local sequences to those retrieved from genbank in the previous step. This function saves all the resulting `fasta` files in two directories, combined sequences in `0.Sequences` and local sequences in `0.AdditionalSequences` (originally downloaded sequences are moved to `0.0.OriginalDownloaded` at this step). Note that `sq.add()` is optional.

-   Fifth, the `sq.curate()` function filters out unreliable sequences based on information listed in genbank (e.g. `PREDICTED`) and on taxonomic information provided by the user. Specifically, this function retrieves taxonomic information from the Global Biodiversity Information Facility (GBIF) database's taxonomic backbone (see alternatives in the advanced vignette to `phruta`). If a given species belongs to a non-target group, this species is dropped from the analyses. This function automatically corrects taxonomy and renames sequences.

-   Sixth, `sq.aln()` performs multiple sequence alignment on `fasta` files. Currently, `phruta` uses the [`DECIPHER` R package](http://www2.decipher.codes/), here. This package allows for adjusting sequence orientation and masking (removing ambiguous sites).

-   Seventh, the `tree.raxml()` function allows users to perform tree inference under `RAxML` for sequences in a given folder. This is a wrapper to `ips::raxml()` and each of the arguments can be customized. The current release of `phruta` can manage both partitioned and unpartitioned analyses. Starting and constrained trees are allowed.

-   Eight, `tree.dating()` enables users to perform time-calibrations of a given phylogeny using `geiger::congruify.phylo()`. `phruta` includes a basic set of comprehensively sampled, time-calibrated phylogenies that are used to extract secondary calibrations for the target phylogeny. Note that sampling in those phylogenies can be examined using `data(SW.phruta)`. Please make sure you have at least **two** groups in common with each of the phylogenies. Similarly, users can choose to run either `PATHd-8` or `treePL`.

## Assembling a molecular dataset for target taxa in phruta

Let's learn how `phruta` works by assembling a molecular dataset at the species level for a few bird clades. Note that athough this tutorial is based on a particular set of taxa, users can decide on what can choose their target clades in other families, orders, or even kingdoms.

Here we will focus on assembling a phylogeny for the new world Quails. Species in this group are classified in the family Odontophoridae, a clade including nearly 34 extant species classified in 10 genera (\ref(fig:quail)). In general, the higher-level taxonomic information in GenBank for the Odontophoridae is largely congruent with recent studies on the systematics of the group. Howeger, GenBank's the taxonomic backbonone classifies _Ptilopachus_, a genus commonly included under the Odontophoridae, as part of the Phasianidae. We will follow more recent studies suggesting that _Ptilopachus_ is nested within the Odontophoridae. As our outgroup, we will select the Phasianidae. Within this clade, we will explicitly focus on the genus _Polyplectron_, a clade of eight extant species. Finally, given that the systematics of the Odontophoridae has been discussed before using morphological and molecular evidence, we will be able to compare the topology of our tree relative to recent studies (Crowe et al. 2006a, b; Cohen et al. 2012; Hosner et al. 2015). 

```{r, eval = TRUE, warning = TRUE, message = TRUE, progress = TRUE}
library(phruta)
```

```{r setup2, eval = TRUE, echo=FALSE, include = FALSE}
assign(".testMode", TRUE, envir = phruta:::pkg.env)
start_time <- Sys.time()
```

So far, we have decided the taxonomic makeup of our analyses. From this point, we could simply check the genetic sampling of previous studies and search for those genes in GenBank for the target taxa (Crowe et al. 2006a, b; Cohen et al. 2012; Hosner et al. 2015). For instance, [review sampling in each of those]. We could use these gene names to assemble a molecular dataset for the Odontophoridae and **Polyplectron** in `phruta`. 

Alternatively, we could use `phruta` to figure out what genes are well sampled in GenBank for both the ingroup and outgroup. For simplicity, we will follow the later procedure in this paper implemented in `gene.sampling.retrieve()` function in `phruta`. The resulting `data.frame`, here named `gs.seqs`, will contain the list of full names for genes sampled in GenBank for the target taxa.

```{r results = 'hide'}
gs.seqs <- gene.sampling.retrieve(organism = c("Odontophoridae", 
                                               "Ptilopachus",
                                               "Polyplectron"), 
                                  speciesSampling = TRUE)
```

For the search terms, `phruta` was able to retrieve the names for `r nrow(gs.seqs)` gene regions from GenBank. The frequency estimates per gene are based on inter-specific sampling (\ref(tab:topGenes)). Note that the `gene.sampling.retrieve()` function provides an estimate of the number of species in GenBank that match the taxonomic criteria of the search term and that have sequences for a given gene region. However, this estimate is only as good as the annotations for genes deposited in GenBank.

We will now generate a preliminary summary of the accession numbers retrieved for the combination of target taxa and gene regions. I call this a preliminary dataset because not all these accession numbers are expected to be in the final molecular dataset. For instance, some sequences may be removed after taxonomic synonyms are identified in the dataset.

Next, we will assemble a species-level summary of accession numbers using the `acc.table.retrieve()` function in `phruta` (i.e. `speciesLevel = TRUE` argument). For simplicity, this tutorial will focus on analyzing gene regions that are sampled in \>20% of the species (`targetGenes` data.frame). The `acc.table` object created below is a `data.frame` object that will later be used to download the relevant gene sequences from GenBank (\ref(tab:AccN)). 

```{r  results = 'hide'}
targetGenes <- gs.seqs[gs.seqs$PercentOfSampledSpecies > 20,]
acc.table <- acc.table.retrieve(
            clades  = c("Odontophoridae", "Ptilopachus", "Polyplectron"),
            genes = targetGenes$Gene,
            speciesLevel = TRUE
          )
```

Since we are interested in retrieving sequences from GenBank using an existing preliminary accession numbers table, we will use the `sq.retrieve.indirect()` function in `phruta`. Please note that there are two versions of `sq.retrieve.*` in `phruta` . The one that we're using in this tutorial, `sq.retrieve.indirect()`, retrieves sequences "indirectly" because it requires a table of accession numbers to be pass as an argument (see the `acc.table.retrieve()` function above). I present the information in this tutorial using `sq.retrieve.indirect()` instead of `sq.retrieve.direct()` because `sq.retrieve.indirect()` is simply more flexible and robust to mistakes. Specifically, `sq.retrieve.indirect()` allow users to correct issues *prior* to downloading/retrieving the sequences. For instance, you can add new sequences, species, populations to the resulting `data.frame` from `acc.table.retrieve()`. You could even manually assemble your own dataset of accession numbers to be retrieved using `sq.retrieve.indirect()`. Instead, `sq.retrieve.direct()` does its best to directly retrieve GenBank sequences for a target set of taxa and set of gene regions. In short, you should be able to catch errors using `sq.retrieve.indirect()` but mistakes will be harder to spot and fix if you're using `sq.retrieve.direct()`. 

We still need to retrieve all the sequences from the accessions table generated using `acc.table`. Note that since we have specified `download.sqs = FALSE` in `sq.retrieve.indirect`, the sequences retrieved from GenBank are returned in a list. If we decide to download the sequences to our working directory using `download.sqs = TRUE`, `phruta` will write all the resulting `fasta` files into a newly created folder `0.Sequences` located in our working directory. 

```{r results = 'hide'}
sqs.downloaded <- sq.retrieve.indirect(acc.table = acc.table, 
                                       download.sqs = FALSE)
```

Now, let's make sure that we are only including sequences that are reliable and from species that we are actually interested in analyzing. We are going to use the `sq.curate()` function for this. We will provide a list of taxonomic names to filter out incorrect sequences (`filterTaxonomicCriteria` argument). For instance, we could simply provide a vector of the genera that we are interested in analyzing. This vector must have a length of `1`, with all the target genera being separated with `|` (e.g. `"Callipepla|Colinus|Dendrortyx"` if we were interested in only those three genera). For now, we will assume that all of the species we downloaded are relevant to the analyses (i.e. `filterTaxonomicCriteria = [AZ]`). Finally, since we are not downloading anything to our working directory, we need to pass our downloaded sequences (`sqs.downloaded` object generated above using the `sq.retrieve.indirect()` function) to the `sqs.object` argument in `sq.curate()`.

```{r results = 'hide'}
sqs.curated <- sq.curate(filterTaxonomicCriteria = '[AZ]',
                         kingdom = 'animals', 
                         sqs.object = sqs.downloaded,
                         removeOutliers = FALSE)
```

Running the `sq.curate()` function will create an object of class `list` (i.e. `sqs.curated`) that includes (1) the curated sequences with original names, (2) the curated sequences with species-level names (`renamed_*` prefix), (3) the accession numbers table (`AccessionTable`; \ref(tab:tw)), and (4) a summary of taxonomic information for all the species sampled in the files (\ref(tab:tw2), \ref(tab:tw3)). From here, we will align the sequences that we just curated using `sq.aln()` with default parameters. We're again passing the output from `sq.curate()`, `sqs.curated`, using the `sqs.object` argument in `sq.aln()`.

```{r results = 'hide'}
sqs.aln <- sq.aln(sqs.object = sqs.curated)
```

The resulting multiple sequence alignments will be saved to the `sqs.aln` object, a list of alignments. For each of the gene regions, we will have access to the original alignment (`Aln.Original`), the masked one (`Aln.Masked`), and information on the masking process. The raw and masked alignments are presented in \ref(fig:alnraw) and \ref(fig:alncur), respectively.

```{r echo = FALSE}
library(ape)
pdf("figs/UF.Raw.pdf", width = 8, height = 8)
par(mfrow = c(ceiling(length(sqs.aln)/2),2))
for (i in seq_along(sqs.aln)) {
  if (!is.null(sqs.aln[[i]]$Aln.Original)) {
  image(sqs.aln[[i]]$Aln.Original,
        main = names(sqs.aln)[i], 
        show.labels = FALSE, 
        xlab = "Position",
        ylab = "Species", 
        legend = FALSE)
  }
}
sv <- dev.off()
```


```{r echo = FALSE}
pdf("figs/UF.Cur.pdf", width = 8, height = 8)
par(mfrow = c(ceiling(length(sqs.aln)/2),2))
for (i in seq_along(sqs.aln)) {
  if (!is.null(sqs.aln[[i]]$Aln.Masked)) {
  image(sqs.aln[[i]]$Aln.Masked, 
        main = names(sqs.aln)[i], 
        show.labels = FALSE, 
        xlab = "Position",
        ylab = "Species", legend = FALSE
        )
  }
}
sv <- dev.off()
```

```{r echo = FALSE}
end_time <- Sys.time()
```

In total, code in this section took `r round(end_time-start_time)` minutes to run in my local machine (RAM: `r get_ram()`, CPU: `r get_cpu()$model_name`, cores: `r get_cpu()$no_of_cores`, platform: `r get_r_version()$platform`, R: `r get_r_version()$version.string`). The performance of `phruta` is largely affected by the complexity of the dataset that is being assembled. On top of that, it is important to remember that multiple functions in `phruta` are directly retrieving data from different websites.


## Basic phylogenetics with `phruta`

### Phylogenetic inference with `phruta` and `RAxML`

Phylogenetic inference in `phruta` is conducted using the `tree.raxml()` function. To use this function, we will have to export our sequence alignments locally. We will follow the same folder structure as if we were exporting everything locally ([NEED FIGURE]). Specifically, our sequence alignments will located in `2.Alignments` and we will exclusively export the alignments that were masked.


```{r message=FALSE, warning=FALSE, include=FALSE}
dir.create("2.Alignments")
lapply(seq_along(sqs.aln), function(x){
  ape::write.FASTA(sqs.aln[[x]]$Aln.Masked, 
                   file = paste0(
                     "2.Alignments/Masked_", names(sqs.aln)[x], ".fasta"
                     )
                   )
})
```

We are now ready to run RAxML. Note that in `tree.raxml()`, we will need to indicate where the aligned sequences are located (`folder` argument), the patterns of the files in the same folder (`FilePatterns` argument; "`Masked_`" in our case), and the total of bootstrap replicates. The `outgroup` argument is optional but since we are interested in calibrating our tree afterwards, we will define it using all the species in _Polyplectron_.


```{r message=FALSE, warning=FALSE}
outgroup <- sqs.curated$Taxonomy[sqs.curated$Taxonomy$genus == 'Polyplectron',]

tree.raxml(folder = '2.Alignments', 
           FilePatterns = 'Masked_', 
           raxml_exec = 'raxmlHPC', 
           Bootstrap = 100,
           outgroup = paste(outgroup$species_names, collapse = ",")
           )
```

The resulting trees are saved in the `3.Phylogeny` folder. For many, the bipartitions tree generated in these runs, `RAxML_bipartitions.phruta`, will be the most relevant. Additionally, `3.Phylogeny` includes `RAxML`-related input and output files. The resulting phylogeny from these analyses is presented in \ref(fig:raxmlphylo).


```{r raxmlplot, message=FALSE, warning=FALSE, include = FALSE}
library(phytools)
raxmlTree <- read.tree("3.Phylogeny/RAxML_bipartitions.phruta")
raxmlTree$root.edge <- 0.01
raxmlTree <- rootedge.to.singleton(raxmlTree)

pdf("figs/raxml_ingroup.pdf", width = 8, height = 8)
plotTree(ladderize(raxmlTree, right = FALSE))
nodelabels(text = raxmlTree$node.label,
           node = 1:raxmlTree$Nnode + Ntip(raxmlTree),
           frame = "none", 
           adj = c(1.1, -0.4)
           )
dev.off()
```


Users can also run partitioned analyses in `RAxML` within `phruta`. This approach is implemented by setting the `partitioned` argument in `tree.raxml` to `TRUE`. For now, partitions are based on the genes are being analyzed. The same model is used to analyze each partition. More details on partitioned analyses can be customized by passing arguments in `ips::raxml`. 

```{r eval = FALSE}
tree.raxml(folder = "2.Alignments", FilePatterns = "Masked_",
           raxml_exec = "raxmlHPC", Bootstrap = 100,
           outgroup = paste(outgroup$species_names, collapse = ","),
           partitioned = TRUE
           )
```


```{r raxmlplotpartitioned, message=FALSE, warning=FALSE, include = FALSE}
library(phytools)
raxmlTree <- read.tree("3.Phylogeny/RAxML_bipartitions.phruta")
raxmlTree$root.edge <- 0.01
raxmlTree <- rootedge.to.singleton(raxmlTree)

pdf("figs/raxml_ingroup_partitioned.pdf", width = 8, height = 8)
plotTree(ladderize(raxmlTree, right = FALSE))
nodelabels(text = raxmlTree$node.label,
           node = 1:raxmlTree$Nnode + Ntip(raxmlTree),
           frame = "none", 
           adj = c(1.1, -0.4)
           )
dev.off()
```



Finally, we note that users will sometimes need to generate constrain tress to seed their searches or limit the tree sampling space. These constrains can be generated using other software including [`TACT: Taxonomic Addition for Complete Trees`](https://github.com/jonchang/tact). 


### Tree dating in `phruta`

Finally, let's perform tree dating in our phylogeny using secondary calibrations extracted from [Scholl and Wiens (2016)](https://royalsocietypublishing.org/doi/pdf/10.1098/rspb.2016.1334). I am only using this study because it has a large phylogeny but I expect to replace it in the near future. Note that `tree.dating()` requieres the user to specify where the `1.Taxonomy.csv` file is. This file is created automatically when sequences are curated using `sq.curate()` and results are exported to your local repository. However, since we were keeping results in the environment, we will have to export it before we can move forward.

```{r message=FALSE, warning=FALSE}
dir.create("1.CuratedSequences")
write.csv(sqs.curated$Taxonomy, '1.CuratedSequences/1.Taxonomy.csv')
```

Tree dating is performed using the `tree.dating()` function in `phruta`. We have to provide the name of the folder containing the `1.Taxonomy.csv` file created in `sq.curate()`. We also have to indicate the name of the folder containing the `RAxML_bipartitions.phruta` file. We will scale our phylogeny using `treePL`. 

```{r message=FALSE, warning=FALSE}
tree.dating(taxonomyFolder = "1.CuratedSequences", 
            phylogenyFolder = "3.Phylogeny", 
            scale = 'treePL')
```

```{r timetree, include=FALSE}
library(strap)
calTree <- read.tree("4.Timetree/family-levelCalibration.tre")
calTree$root.edge <- 1
calTree2 <- calTree
calTree2$root.time <- max(nodeHeights(calTree2))

#Need to adjust margins
pdf("figs/phylo_ingroup.pdf", width = 8, height = 8)
geoscalePhylo(tree = ladderize(calTree2,right = FALSE), 
              units = c("Epoch", "Age"), 
              boxes = "Epoch",   
              cex.tip = 0.7, 
              cex.age = 0.7, 
              cex.ts = 0.7, 
              label.offset = 0, 
              x.lim = c(-15, max(nodeHeights(calTree2))), 
              lwd = 3, 
              width = 2, 
              quat.rm = TRUE, 
              root.edge = TRUE)
dev.off()
```

Running this line will result in a new folder `4.Timetree`, including the different time-calibrated phylogenies obtained (if any) and associated secondary calibrations used in the analyses. The resulting time-calibrated tree is presented in \ref(fig:timecaltree).


## Advanced methods with `phruta` 

### Curating taxonomic names

You can use `taxonomy.retrieve()`, a function implemented inside `sq.curate()` in `phruta` to curate species names regardless of the kingdom. For instance, the block of code below will curate taxonomic names using the gbif taxonomic backbone. Note that the `kingdom` argument in `taxonomy.retrieve()` can be set to `NULL`, meaning that there wont be indication on the kingdom when performing taxonomic searches.


```{r message=FALSE, warning=FALSE}
phruta:::taxonomy.retrieve(species_names=c("Felis_catus", "PREDICTED:_Vulpes",
                  "Phoca_largha", "PREDICTED:_Phoca" ,
                  "PREDICTED:_Manis" , "Felis_silvestris" , "Felis_nigripes"),
                  database = 'gbif')
```

However, `gbif` is efficient for retrieving accurate taxonomy when we provide details on the `kingdom`. Given that all the species we're interested in are animals, we could just use the following block of code to curate taxonomic names.

```{r message=FALSE, warning=FALSE}
phruta:::taxonomy.retrieve(species_names = c("Felis_catus", "PREDICTED:_Vulpes",
                  "Phoca_largha", "PREDICTED:_Phoca" ,
                  "PREDICTED:_Manis" , "Felis_silvestris" , "Felis_nigripes"),
                  database = 'gbif', kingdom = 'animals')
```

Depending on your sampling, you could also do the same for plants by using `plants` in the `kingdom` argument instead of `animals`. Now, what if we were interested in following other databases to retrieve taxonomic information for the species in our database? The latest version of `phruta` allow users to select the desired database. The databases follow the `taxize::classification()` function. Options are: `ncbi`, `itis`, `eol`, `tropicos`, `nbn`, `worms`, `natserv`, `bold`, `wiki`, and `pow`. Please select only one. Note that the `gbif` option in `taxize::classification()` is replaced by the internal `gbif` in `phruta`.


```{r message=FALSE, warning=FALSE, eval = FALSE}
phruta:::taxonomy.retrieve(species_names = c("Felis_catus", "PREDICTED:_Vulpes",
                  "Phoca_largha", "PREDICTED:_Phoca" ,
                  "PREDICTED:_Manis" , "Felis_silvestris" , "Felis_nigripes"),
                  database = 'itis')
```


### Running PartitionFinder in `phruta`

With the current version of `phruta`, users are able to run `PartitionFinder` v1 within `R`. For this, users should provide the name of the folder where the alignments are stored, a particular pattern in the file names (`Masked_` in our case), and which models will be run in `PartitionFinder`. This function will download `PartitionFinder`, generate the input files, and run it all within R. The output files will be in a new folder within the working directory.

```{r eval=FALSE}
sq.partitionfinderv1(folderAlignments = "2.Alignments",
                    FilePatterns = "Masked_",
                    models = "all"
                    )
```

Unfortunately, the output files are not integrated with the current `phruta` pipeline. This will be part of a new release. However, users can still perform gene-based partitioned analyses within `RAxML` or can use `PartitionFinder`'s output files to inform their own analyses outside `phruta`.


### Identifying rogue taxa

`phruta` can help users run `RogueNaRok` implemented in the `Rogue` R package. Users can then examine whether rogue taxa should be excluded from the analyses. `tree.roguetaxa()` uses the bootstrap trees generated using the `tree.raxml()` function along with the associated best tree to identify rogue taxa. Rogue taxa for the analyses presented in this paper are shown in \ref(table:rogue)

```{r message=FALSE, warning=FALSE}
tree.roguetaxa(folder = "3.Phylogeny")
```


## Reproducibility with `phruta` 

One of the central points of developing `phruta` was related to increasing the reproducibility of simple phylogenetic analyses. By either compiling or calling alternative tools that are commonly used to assemble species-level molecular and phylogenetic datasets inside an `R` package, `phruta` allows users to generate a clear, structured, and reproducible workflow. In fact, `phruta` is conceived as a package that allow users to choose between at least two alternative options to share their workflow. First, users can simply provide access to their workflow in an `R` script. For instance, this file can be stored in GitHub along with all the intermediate files that are created each each given step. Alternatively, given that the information in databases is constantly changing, users can share their `R` script and associated workspace to assure that the versions of the retrieved files correspond to specific versions of the databases. 

As an example, all the analyses and sequences analyzed in the current manuscript will be available to anyone in a `RMarkdown` file accompanied by a n `R` workspace `phruta_ms.RData`. The `RMarkdown` will allow users to replicate the analyses presented in the current paper. With the workspace, users can choose to use the stored objects instead of performing searches on GenBank and additional taxonomic databases. These two files confer full reproducibility to the analyses presented in this article.

```{r}
save.image(file = "phruta_ms.RData")
```


## Performance

Assessing the performance of `phruta` is intrinsically challenging However, it is expected for functions that scrape and curate information stored in GenBank to show an slow performance on search terms (e.g. clades) with extensive genetic sampling. Below, I focus on comparing the distribution of durations for assembling molecular datasets in seven clades with different sample size.  These estimates encompass all the steps outlined above between `gene.sampling.retrieve()` and `sq.curate`. Time estimates were estimated using the `microbenchmark` `R` package. \ref(fig:benchmark) shows a summary of `phruta` performance in different clades.



```{r include=FALSE}
library(microbenchmark)

if ( !"Bench.csv" %in% list.files()) {

clades <- c("Odontophoridae", 
            "Trichostrongylidae", 
            "Dicksoniaceae", 
            "Halobates",
            "Ctenidae",
            "Psocidae",
            "Treubiales"
            ) 

testBench <- function(clade){
gs.seqs <- gene.sampling.retrieve(organism = clade, 
                                  speciesSampling = TRUE)

targetGenes <- gs.seqs[gs.seqs$PercentOfSampledSpecies > 20,]
acc.table <- acc.table.retrieve(
  clades  = clade,
  genes = targetGenes$Gene,
  species = NULL,
  speciesLevel = TRUE
)

sqs.downloaded <- sq.retrieve.indirect(acc.table = acc.table, 
                                       download.sqs = FALSE)

sqs.curated <- sq.curate(filterTaxonomicCriteria = '[AZ]',
                         kingdom = NULL, 
                         sqs.object = sqs.downloaded,
                         removeOutliers = FALSE)

}

benchResults <- lapply(clades, function(x){
  summary(microbenchmark(testBench(x), times = 5))
})

BenchComplete <- do.call(rbind, benchResults)
BenchComplete <- cbind.data.frame(clade = clades, BenchComplete)
write.csv(BenchComplete, "Bench.csv")
}

```


```{r include=FALSE}
BenchData <- read.csv("Bench.csv")
BenchData$sd <- ((BenchData$uq - BenchData$lq)/2)/BenchData$neval

BenchData$clade <- paste0(BenchData$clade, 
                          "\nSpecies = ", BenchData$n, 
                          "\nHits = ",  BenchData$Hits )

library(ggplot2)
p <- ggplot(BenchData, aes(x = reorder(clade, mean), y = mean, fill = clade)) + 
  geom_bar(stat = "identity", color = "black", 
           position = position_dodge()) +
  geom_errorbar(aes(ymin = mean - sd, ymax = mean + sd), width = .2,
                 position = position_dodge(.9))  +
  geom_hline(yintercept = c(60, 120, 180), lty = 'dashed') +
  theme_bw()+ 
  theme(legend.position = "none")

pdf("figs/benchmark.pdf", width = 8, height = 8)
print(p)
dev.off()

```



```{=html}

# References
\setlength{\parindent}{-0.2in}
\setlength{\leftskip}{0.2in}
\setlength{\parskip}{8pt}
\vspace*{-0.2in}
\noindent

```


\newpage

```{r quail, include=TRUE, fig.cap=c("Quail placeholder. Phyto by Brent Myers"), echo=FALSE}
knitr::include_graphics("./figs/Quail.jpg")
```


\newpage

```{r alnraw, include=TRUE, fig.cap=c("Raw alignments for gene regions sampled in more than 20% of the species in GenBank"), echo=FALSE}
knitr::include_graphics("./figs/UF.Raw.pdf")
```


\newpage
```{r alncur, include=TRUE, fig.cap=c("Curated alignments for gene regions sampled in more than 20% of the species in GenBank"), echo=FALSE}
knitr::include_graphics("./figs/UF.Cur.pdf")
```

\newpage

```{r raxmlphylo, include=TRUE, fig.cap=c("RAxML phylo"), echo=FALSE}
knitr::include_graphics("./figs/raxml_ingroup.pdf")
```

\newpage

```{r timecaltree, include=TRUE, fig.cap=c("Time-calibrated phylogeny phylo"), echo=FALSE}
knitr::include_graphics("./figs/phylo_ingroup.pdf")
```

\newpage


```{r benchmark, include=TRUE, fig.cap=c("Time-calibrated phylogeny phylo"), echo=FALSE}
knitr::include_graphics("./figs/benchmark.pdf")
```



\newpage

```{r topGenes, echo=FALSE}
kbl(head(gs.seqs), longtable = TRUE, booktabs = TRUE, caption = "Top six genes sampled in GenBank  for species in Odontophoridae and Polyplectron.", format = "latex") %>%
  kable_styling(latex_options = c("hold_position", "repeat_header"))
```

\newpage

```{r AccN, echo=FALSE}
row.names(acc.table) <- NULL
kbl(acc.table[,-2], longtable = TRUE, booktabs = TRUE, caption = 'Summary of potential accession numbers for the species in Odontophoridae, our ingroup, and Polyplectron, outgroup genus. This list of sequences has not been curated yet.', format = "latex") %>%
  kable_styling(latex_options = c("hold_position", "repeat_header"))
```

\newpage

```{r tw, echo = FALSE}
kbl(sqs.curated$AccessionTable[,-5], longtable = TRUE, booktabs = TRUE, caption = "Accession numbers for the retrieved sequences. This dataset has been curated.", format = "latex") %>%
  kable_styling(latex_options = c("hold_position", "repeat_header"))%>%
landscape()
```

\newpage

```{r tw2, echo = FALSE}
kbl(sqs.curated$Taxonomy[,c(1:7)], longtable = TRUE, booktabs = TRUE, caption = "Taxonomic information for the retrieved species", format = "latex") %>%
  kable_styling(latex_options = c("hold_position", "repeat_header"))%>%
landscape()
```

\newpage

```{r tw3, echo = FALSE}
kbl(sqs.curated$Taxonomy.Sampling[,c(1:7)], longtable = TRUE, booktabs = TRUE, caption = "Taxonomic sampling across gene regions", format = "latex") %>%
  kable_styling(latex_options = c("hold_position", "repeat_header"))%>%
landscape()
```

\newpage

```{r include=FALSE}
rt <- read.csv("3.1.RogueTaxa/RogueTaxa.csv")[,-1]
```


```{r rogue, echo = FALSE}
kbl(rt, caption = "Results of RogueNaRock", longtable = TRUE, booktabs = TRUE, format = "latex") %>%
  kable_styling(latex_options = c("hold_position", "repeat_header")) %>%
  landscape()
```
\newpage


```{r include=FALSE}
unlink(c("0.Sequences", 
         "1.CuratedSequences", 
         "2.Alignments", 
         "2.1.PartitionFinderv1",
         "3.Phylogeny", 
         "3.1.RogueTaxa",
         "3.2.Phylogeny.constraint",
         "4.Timetree",
         "_cache"), 
       recursive = TRUE)
```



